pkg/controller/master/addon/addon.go

```diff
@@ -0,0 +1,301 @@
+	job   ctlcorev1.JobController
+}
+
+var (
+	defaultDeployDuration = &metav1.Duration{Duration: 10 * time.Minute}
+)
+
+const (
+	defaultValuesAnnotation = "harvesterhci.io/addon-defaults"
+)
+
+func Register(ctx context.Context, management *config.Management, opts config.Options) error {
+	addonController := management.HarvesterFactory.Harvesterhci().V1beta1().Addon()
+	helmController := management.HelmFactory.Helm().V1().HelmChart()
+	jobController := management.BatchFactory.Batch().V1().Job()
+	h := &Handler{
+		helm:  helmController,
+		addon: addonController,
+		job:   jobController,
+	}
+
+	addonController.OnChange(ctx, "deploy-addon", h.OnAddonChange)
+	addonController.OnChange(ctx, "monitor-addon", h.MonitorAddon)
+	addonController.OnChange(ctx, "monitor-changes", h.MonitorChanges)
+	relatedresource.Watch(ctx, "watch-helmcharts", h.ReconcileHelmChartOwners, addonController, helmController)
+	return nil
+}
+
+// MonitorChanges will trigger updates to helm chart based on chages in values
+func (h *Handler) MonitorChanges(key string, a *harvesterv1.Addon) (*harvesterv1.Addon, error) {
+	if a == nil || !a.DeletionTimestamp.IsZero() {
+		return nil, nil
+	}
+
+	if a.Status.Status != harvesterv1.AddonDeployed && a.Status.Status != harvesterv1.AddonFailed {
+		return a, nil
+	}
+
+	hc, err := h.helm.Get(a.Namespace, a.Name, metav1.GetOptions{})
+
+	var redeployChart bool
+	if err != nil {
+		// HelmChart is missing. Reset Addon to allow normal reconcile
+		if apierrors.IsNotFound(err) {
+			redeployChart = true
+		} else {
+			return a, fmt.Errorf("error fetching helm chart during monitor changes: %v", err)
+		}
+	}
+
+	// if chart is being deleted then but addon is not then redeploy chart
+	// this is possible since helm-controller adds a finalizer to uninstall the chart when HelmChart CRD is being deleted
+
+	if hc.DeletionTimestamp != nil {
+		redeployChart = true
+	}
+
+	if redeployChart {
+		a.Status.Status = ""
+		return h.addon.UpdateStatus(a)
+	}
+
+	vals, err := defaultValues(a)
+	if err != nil {
+		return a, fmt.Errorf("error generating default values during monitor changes: %v", err)
+	}
+	if hc.Spec.ValuesContent != vals || hc.Spec.Version != a.Spec.Version || hc.Spec.Chart != a.Spec.Chart || hc.Spec.Repo != a.Spec.Repo {
+		hc.Spec.Chart = a.Spec.Chart
+		hc.Spec.Repo = a.Spec.Repo
+		hc.Spec.Version = a.Spec.Version
+		hc.Spec.ValuesContent = vals
+		_, err = h.helm.Update(hc)
+		if err != nil {
+			return a, fmt.Errorf("error updating helm chart during monitor changes: %v", err)
+		}
+	}
+
+	// reset status to AddonEnable to trigger reconcile for job
+	a.Status.Status = harvesterv1.AddonEnabled
+	return h.addon.UpdateStatus(a)
+}
+
+// MonitorAddon will track the deployment of Addon
+func (h *Handler) MonitorAddon(key string, a *harvesterv1.Addon) (*harvesterv1.Addon, error) {
+	if a == nil || !a.DeletionTimestamp.IsZero() {
+		return nil, nil
+	}
+
+	if a.Status.Status != harvesterv1.AddonEnabled {
```

Comment:
may also return if the Addon is deployed successfully?

Reply:
Ideally this would only be called for the first deploy. Any subsequent changes to Addon/Helm chart would be via the MonitorChanges handler